### PR TITLE
Implement SetQueryParams in the Provider Interface

### LINKED ISSUE #13: feature: Providers must implement Query Parameters
# Feature Request

## Summary

Include `SetQueryParams()` method in the Provider interface implementation

## Motivation

As of now, QP is passed as a parameter to `FetchData()` method to implement `Provider` interface. However, this mandates passing QP as an argument for every `FetchData()` call for that Provider client. 
- As the QPs aren't modified frequently for a provider, it would reduce computational space and time when attached to the client struct just once.
- Should the QPs conditionally change, or another use case calls for adhoc changes, a new Provider client may be created with modified QP map.
- Makes the code cleaner, and pragmatic.

## Detailed Design

In `providers` package:
type Provider interface {
	~FetchData(queryParams map[string]string) (*plumber.BaseData, error)~
	FetchData() (*plumber.BaseData, error)
	SetQueryParams(qp map[string]string)
}

In `open_meteo.go`, func `OpenMeteoQueryParams()` must be a method on `OpenMeteoProvider` type.

Similar implementation for the upcoming meteoblue implementation as well.

## Alternatives

Retain the existing design of having independent functions to set QP rather than mandating in the Provider interface.

## Additional Context

None.

## Acceptance Criteria

- [ ] Remove the `queryParams map[string]string` parameter from `FetchData()` method `Provider` interface.
- [ ] Include `SetQueryParams()` method to the `Provider` interface.
- [ ] Make appropriate changes to open-meteo provider to implement the updated `Provider` interface.

## Additional Information

#12 #5

### PR DESCRIPTION
**Pull request by yendelevium**
- Changed the function signature of the FetchData method in the Provides interface
- Added the SetQueryParams method to the providers aligned to the concerns raised in #13

Resolves #13
References #5

### CODE REVIEW

File: providers/meteoblue.go [hunk 1]
```diff
@@ -96,4 +104,21 @@ func (p *MeteoBlueProvider) FetchData(qp map[string]string) (*plumber.BaseData,
 	return baseData, nil
 }
 
-// Maybe add a MeteoBlueQueryParams like OpenMeteoQueryParams to have some standard for all providers?
+// Set the QueryParams for the request
+func (p *MeteoBlueProvider) SetQueryParams(coords *plumber.Coordinates) map[string]string {
+	// Get the config to accesss the API Key
```
Comment (JrGkOG) [inline providers/meteoblue.go]: we can add a check to see whether the coords aren't nil ?

**Review by raamsri — CHANGES_REQUESTED** [review]
1. This patch makes FetchData() receive and set Location dynamically. That's good.

2. The idea is to reduce the movement/redundant passing of QP, however this patch hasn't addressed it yet; since it returns the QP data, it must then be passed around. Please consider the following changes?

a. `OpenMeteoProvider` struct must have `QueryParams` field to hold the QP values that is of type map[string]string

b. `SetQueryParams(*plumber.Coordinates)` must set the field OpenMeteoProvider.QueryParams with appropriate values. Since it's a method on `OpenMeteoProvider` that sets its own field, it doesn't have to return anything. In this patch, it's instead returning the query params rather than setting it.

Please correct if I have misunderstand something.

c. `NewOpenMeteoProvider()` and `NewMeteoBlueProvider()` must appropriately call their respective SetQueryParams() method to set the QP before returning a new struct. (A default location, or nil, may be set here.) It would be convenient to also make the appropriate `p.client.SetQueryParams()` to resty with this QP before returning the struct, so that it sticks and only the location cords may have to passed while fetching data. 

d. `FetchData()` need not change location data of `QueryParams` field of the MeteoProvider struct as it's a transient call. This side effect is misleading, and not necessary. It simply has to set the passed location co-ords via `p.client.SetQueryParams()` as the other QP is already set earlier while creating a new provider. That way, only the location co-ords has to be set freshly, and that's the intention as well.

**Comment by yendelevium** [comment]
> 1. This patch makes FetchData() receive and set Location dynamically. That's good.
> 2. The idea is to reduce the movement/redundant passing of QP, however this patch hasn't addressed it yet; since it returns the QP data, it must then be passed around. Please consider the following changes?
> 
> a. `OpenMeteoProvider` struct must have `QueryParams` field to hold the QP values that is of type map[string]string
> 
> b. `SetQueryParams(*plumber.Coordinates)` must set the field OpenMeteoProvider.QueryParams with appropriate values. Since it's a method on `OpenMeteoProvider` that sets its own field, it doesn't have to return anything. In this patch, it's instead returning the query params rather than setting it.
> 
> Please correct if I have misunderstand something.
> 
> c. `NewOpenMeteoProvider()` and `NewMeteoBlueProvider()` must appropriately call their respective SetQueryParams() method to set the QP before returning a new struct. (A default location, or nil, may be set here.) It would be convenient to also make the appropriate `p.client.SetQueryParams()` to resty with this QP before returning the struct, so that it sticks and only the location cords may have to passed while fetching data.
> 
> d. `FetchData()` need not change location data of `QueryParams` field of the MeteoProvider struct as it's a transient call. This side effect is misleading, and not necessary. It simply has to set the passed location co-ords via `p.client.SetQueryParams()` as the other QP is already set earlier while creating a new provider. That way, only the location co-ords has to be set freshly, and that's the intention as well.

Makes sense, I will implement them.

File: providers/meteoblue.go (hunk 1 above)
Comment (yendelevium) [reply providers/meteoblue.go]: I thought we could do this when we define the endpoints, before calling fetchData itself. If it's nil, we don't make a request to the provider and write the err response

**Comment by yendelevium** [comment]
Have implemented the changes @adofm @raamsri , Please review

File: providers/meteoblue.go [hunk 2]
```diff
@@ -45,16 +46,26 @@ func NewMeteoBlueProvider() (*MeteoBlueProvider, error) {
...
 }
 
-func (p *MeteoBlueProvider) FetchData(qp map[string]string) (*plumber.BaseData, error) {
-	// Creating a new request everytime we fecth data
-	p.client.NewRequest()
-	resp, err := p.client.SetQueryParams(qp).Get(p.config.APIPath)
+func (p *MeteoBlueProvider) FetchData(coords *plumber.Coordinates) (*plumber.BaseData, error) {
+	resp, err := p.client.
+		SetQueryParams(map[string]string{
+			"latitude":  fmt.Sprintf("%f", coords.Latitude),
```
Comment (raamsri) [inline providers/meteoblue.go]: "lat"

File: providers/open_meteo.go [hunk 3]
```diff
@@ -46,16 +47,27 @@ func NewOpenMeteoProvider() (*OpenMeteoProvider, error) {
...
 }
 
 // FetchData fetches API data from open-meteo provider for the given query parameters map
-func (p *OpenMeteoProvider) FetchData(qp map[string]string) (*plumber.BaseData, error) {
-	p.client.NewRequest()
-	resp, err := p.client.SetQueryParams(qp).Get(p.config.APIPath)
+func (p *OpenMeteoProvider) FetchData(coords *plumber.Coordinates) (*plumber.BaseData, error) {
+	resp, err := p.client.
+		SetQueryParams(map[string]string{
+			"latitude":  fmt.Sprintf("%f", coords.Latitude),
```
Comment (raamsri) [inline providers/open_meteo.go:66]: "lat" and "long"

**Review by raamsri — APPROVED** [review]
LGTM. ~~Just the typo of the query fields "lat" and "long" needs fixing.~~

File: providers/open_meteo.go (hunk 3 above)
Comment (raamsri) [reply providers/open_meteo.go:66]: Please ignore. What you have put here is correct.

**Comment by raamsri** [comment]
> LGTM. ~Just the typo of the query fields "lat" and "long" needs fixing.~

Open Meteo uses "latitude" and "longitude" while Meteoblue uses "lat" and "lon"

File: providers/meteoblue.go [hunk 4]
```diff
@@ -45,16 +46,26 @@ func NewMeteoBlueProvider() (*MeteoBlueProvider, error) {
...
+			"longitude": fmt.Sprintf("%f", coords.Longitude),
```
Comment (raamsri) [inline providers/meteoblue.go]: Open-Meteo uses "latitude" and "longitude" while Meteoblue uses "lat" and "lon". 

This forms the URL as `"...&lat=0.000000&latitude=10.901838&lon=0.000000&longitude=76.899845&tz=GMT"`

**Review by raamsri — CHANGES_REQUESTED** [review]
There's a mix up between the use of "lat" and "lon" for Meteo Blue, in `SetQueryParams()` call within `FetchData()`.

File: providers/meteoblue.go (hunk 4 above)
Comment (yendelevium) [reply providers/meteoblue.go]: Hmm, then how about we don't set the latitude and longitude / lat - long while setting the default query params and directly set them for the first time in the FetchData() call?

Comment (raamsri) [reply providers/meteoblue.go]: I reckon that should be okay.

Comment (yendelevium) [reply providers/meteoblue.go]: Done

**Review by raamsri — APPROVED** [review]

**Pull request merged**